controller/share_manager_controller.go

```diff
@@ -390,8 +390,8 @@ func (c *ShareManagerController) syncShareManagerPod(sm *longhorn.ShareManager)
 
 	// no active workload, there is no need to keep the share manager around and the volume can be detached
 	if volume.Status.KubernetesStatus.LastPodRefAt != "" || volume.Status.KubernetesStatus.LastPVCRefAt != "" {
-		isVolumeAttached := volume.Status.State == types.VolumeStateAttached || volume.Status.State == types.VolumeStateAttaching
-		if sm.Status.State != types.ShareManagerStateStopped && isVolumeAttached && !volume.Status.FrontendDisabled {
+		isVolumeAttached := volume.Spec.NodeID == sm.Status.OwnerID && !volume.Status.FrontendDisabled
```

Comment:
For 1.2 we really should refactor the volume controller so it can handle it's own detach/attach when a different node is specified in the spec. It doesn't make sense that the client needs to manage the detach/attach statemachine.

The reason I consider FrontedDisable, is so I don't detach volumes that are attached in maintenance mode.

Reply:
For `!volume.Status.FrontendDisabled`, I mean putting this check in line 394 rather than use it as a condition of `isVolumeAttached`

I know we will refactor this part in the following release. But the logic in this function is already becoming more and more complicated. At the first glance, this function seems to check/update sm and create/delete sm pod back and forth.

Reply:
I am fine with doing this right now as long as it's contained in the share manager. We can look deeper into how would we refactor this later.

@shuo-wu can you make sure the logic is sound?

Reply:
I will refactor the volume attachment handling in a separate pr while working on this issue:
https://github.com/longhorn/longhorn/issues/2079

Reply:
The naming of this variable is a little bit weird since `volume.Spec.NodeID == sm.Status.OwnerID` doesn't mean the volume is already attached. But this variable is designed for cleaning up `volume.Spec.NodeID` hence in general it looks fine.

Reply:
I refactored the share-manager-controller to make it easier to understand.
the first commit is only about the node drain scenario.
the second commit contains the full new logic.